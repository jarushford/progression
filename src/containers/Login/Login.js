import React, { Component } from 'react'
import { connect } from 'react-redux'
import { addNewUser, loginUser } from '../../utils/apiCalls'
import { Redirect } from 'react-router-dom'
import { setUser } from '../../actions'

class Login extends Component {
  constructor() {
    super()
    this.state = {
      loginEmail: '',
      loginPassword: '',
      signUpEmail: '',
      signUpName: '',
      signUpPassword: '',
      loggedIn: false
    }
  }

  handleLogin = async (e) => {
    const { setUser } = this.props
    e.preventDefault()
    const user = {
      email: this.state.loginEmail.toLowerCase(),
      password: this.state.loginPassword
    }
    try {
      const currentUser = await loginUser(user)
      setUser(currentUser)
      this.setState({ loggedIn: true })
    } catch(error) {
      console.log(error.message)
    }
  }

  handleSignUp = async (e) => {
    const { setUser } = this.props
    e.preventDefault()
    const user = {
      name: this.state.signUpName,
      email: this.state.signUpEmail,
      password: this.state.signUpPassword
    }
    try {
      await addNewUser(user)
      setUser(user)
      this.setState({ loggedIn: true })
    } catch(error) {
      console.log(error.message)
    }
  }

  handleInput = (e) => {
    const { name, value } = e.target
    this.setState({ [name]: value })
  }

  render() {
    const { loginEmail, loginPassword, signUpName, signUpEmail, signUpPassword, loggedIn } = this.state

    if (loggedIn) {
      return <Redirect to='/' />
    }

    return (
      <section className="login-section">
        <form onSubmit={this.handleLogin}>
          <h1>Log In</h1>
          <input
            name="loginEmail"
            value={loginEmail}
            onChange={this.handleInput}
            placeholder="email"
            type="email"
          />
          <input
            name="loginPassword"
            value={loginPassword}
            onChange={this.handleInput}
            placeholder="password"
            type="password"
          />
          <button>Login</button>
        </form>
        <form onSubmit={this.handleSignUp}>
          <h1>Sign Up</h1>
          <input
            name="signUpName"
            value={signUpName}
            onChange={this.handleInput}
            placeholder="name"
          />
          <input
            name="signUpEmail"
            value={signUpEmail}
            onChange={this.handleInput}
            placeholder="email"
            type="email"
          />
          <input
            name="signUpPassword"
            value={signUpPassword}
            onChange={this.handleInput}
            placeholder="password"
            type="password"
          />
          <button>Sign Up</button>
        </form>
      </section>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  setUser: (user) => dispatch(setUser(user))
})

export default connect(null, mapDispatchToProps)(Login)