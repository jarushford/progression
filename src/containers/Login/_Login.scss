.login-section {
  @include pageBody();
  margin-top: 120px;
  display: flex;
  height: 70vh;

  form {
    width: 100%;
    margin: 20px auto;
    display: flex;
    flex-direction: column;
    align-items: center;

    h1 {
      @include pageTitle();
      color: $accent;
      margin: 40px auto 10px;
    }

    input {
      width: 240px;
      height: 35px;
      margin: 13px auto;
      padding-left: 6px;
      padding-top: 2px;
      font-size: 1.2rem;
      font-weight: 600;
      outline: none;
      border: none;
      border-radius: 2px;

      &::placeholder {
        font-size: 1.2rem;
        color: rgba(150, 150, 150, 0.7);
      }
    }

    button {
      @include button();
      margin-top: 20px;
    }
  }

  form:nth-child(1) {
    border-right: 1px solid $ruleLine;
  }
}

@media screen and (max-width: 750px) {
  .login-section {
    flex-direction: column;
    height: 97vh;

    form h1 {
      margin-top: 20px;
    }

    form:nth-child(1) {
      border-right: none;
      padding-bottom: 10px;
    }
  }
}

@media screen and (max-width: 400px) {
  .login-section {
    margin: 120px 10px 20px;
    height: 105vh;
  }
}

@media screen and (max-width: 360px) {
  .login-section {
    margin: 120px 10px 20px;
    height: 120vh;
  }
}

