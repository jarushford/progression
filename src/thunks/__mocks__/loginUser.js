export const loginUserThunk = jest.fn()