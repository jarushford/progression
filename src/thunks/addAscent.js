import { fetchAscentsThunk } from './fetchAscents'
import { clearAscents, setError } from '../actions'

export const addAscentThunk = (ascent) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/ascents/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(ascent),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add ascent')
      }
      await dispatch(clearAscents())
      dispatch(fetchAscentsThunk(ascent.user_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}