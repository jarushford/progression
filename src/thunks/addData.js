import { clearThunkHelper } from '../utils/thunkHelpers'
import { fetchDataThunk } from './fetchData'
import { setError } from '../actions'

export const addDataThunk = (data, type) => {
  return async (dispatch) => {
    try {
      const url = `http://localhost:3000/api/progressionusers/${type}s/new`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(data),
        headers: {
          "Content-Type": "application/json"
        }
      })

      if (!response.ok) {
        throw Error(`Could not add ${type}`)
      }

      await dispatch(clearThunkHelper(type)())
      type === 'journal' || type === 'milestone'
        ? dispatch(fetchDataThunk(data.user_id, data.project_id, type))
        : dispatch(fetchDataThunk(data.user_id, '', type))

      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}