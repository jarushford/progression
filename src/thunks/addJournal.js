import { fetchJournalThunk } from './fetchJournal'
import { clearJournal, setError } from '../actions'

export const addJournalThunk = (entry) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/journal/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(entry),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add entry')
      }
      await dispatch(clearJournal())
      dispatch(fetchJournalThunk(entry.user_id, entry.project_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}