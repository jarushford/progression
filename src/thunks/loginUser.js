import { setError, setUser } from '../actions'
import { fetchDataThunk } from './fetchData'

export const loginUserThunk = (user) => {
  return async (dispatch) => {
    try {
      
      let root
      if (process.env.NODE_ENV === 'development') {
        root = 'http://localhost:3000'
      } else {
        root = 'https://progression-backend.herokuapp.com'
      }

      const url = `${root}/api/progressionusers`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(user),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if (!response.ok) {
        throw new Error('Incorrect email or password')
      }
      const result = await response.json()
      const currentUser = result.data
      dispatch(setUser(currentUser))
      await dispatch(fetchDataThunk(currentUser.id, '', 'ascent'))
      await dispatch(fetchDataThunk(currentUser.id, '', 'project'))
      await dispatch(fetchDataThunk(currentUser.id, '', 'workout'))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}