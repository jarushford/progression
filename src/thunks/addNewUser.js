import { setError } from '../actions'
import { loginUserThunk } from './loginUser'

export const addNewUserThunk = (user) => {
  return async (dispatch) => {
    try {
      const url = `https://progression-backend.herokuapp.com/api/progressionusers/new`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(user),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw new Error('User already exists, try another email!')
      }
      await dispatch(loginUserThunk(user))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}