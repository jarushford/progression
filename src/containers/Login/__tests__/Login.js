import React from 'react'
import { shallow } from 'enzyme'
import { addNewUserThunk } from '../../../thunks/addNewUser'
import { loginUserThunk } from '../../../thunks/loginUser'
import { mapDispatchToProps, Login } from '../Login'

jest.mock('../../../thunks/addNewUser')
jest.mock('../../../thunks/loginUser')

describe('Login', () => {
  let mockLogin
  let mockSignup

  beforeEach(() => {
    mockLogin = jest.fn().mockImplementation(() => true)
    mockSignup = jest.fn().mockImplementation(() => true)
  })

  describe('Login Component', () => {
    it('should match the snapshot', () => {
      const wrapper = shallow(<Login />)

      expect(wrapper).toMatchSnapshot()
    })

    it('should have the correct default state', () => {
      const expected =  {
        loginEmail: '',
        loginPassword: '',
        signUpEmail: '',
        signUpName: '',
        signUpPassword: '',
        loggedIn: false
      }
      const wrapper = shallow(<Login />)

      expect(wrapper.state()).toEqual(expected)
    })

    it('should login an existing user', async () => {
      const wrapper = shallow(<Login loginUser={mockLogin} />)
      const e = { preventDefault: jest.fn() }
      wrapper.setState({
        loginEmail: 'thegrey@gmail.com',
        loginPassword: 'Shadowfax'
      })

      await wrapper.find('form').first().simulate('submit', e)

      expect(wrapper.state().loggedIn).toEqual(true)
    })

    it('should sign up a new user', async () => {
      const wrapper = shallow(<Login addNewUser={mockSignup} />)
      const e = { preventDefault: jest.fn() }
      wrapper.setState({
        signUpName: 'Gandalf',
        signUpEmail: 'thegrey@gmail.com',
        signUpPassword: 'Shadowfax'
      })

      await wrapper.find('form').last().simulate('submit', e)

      expect(wrapper.state().loggedIn).toEqual(true)
    })

    it('should not reset state if login info is incorrect', async () => {
      mockLogin.mockImplementation(() => false)
      const wrapper = shallow(<Login loginUser={mockLogin} />)
      const e = { preventDefault: jest.fn() }
      wrapper.setState({
        loginEmail: 'thegrey@gmail.com',
        loginPassword: 'Shadowfax'
      })

      await wrapper.find('form').first().simulate('submit', e)

      expect(wrapper.state().loggedIn).toEqual(false)
    })

    it('should not reset state if sign up user exists already', async () => {
      mockSignup.mockImplementation(() => false)
      const wrapper = shallow(<Login addNewUser={mockSignup} />)
      const e = { preventDefault: jest.fn() }
      wrapper.setState({
        signUpName: 'Gandalf',
        signUpEmail: 'thegrey@gmail.com',
        signUpPassword: 'Shadowfax'
      })

      await wrapper.find('form').last().simulate('submit', e)

      expect(wrapper.state().loggedIn).toEqual(false)
    })

    it('should update state on change of the inputs', async () => {
      const wrapper = shallow(<Login />)
      const e = { target: { name: 'loginEmail', value: 'G' } }

      await wrapper.find('input').first().simulate('change', e)

      expect(wrapper.state().loginEmail).toEqual('G')
    })
  })

  describe('mapDispatchToProps', () => {
    it('should return a props object with a method loginUser', () => {
      const mockDispatch = jest.fn()
      const expected = loginUserThunk()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.loginUser()

      expect(mockDispatch).toBeCalledWith(expected)
    })

    it('should return a props object with a method addNewUser', () => {
      const mockDispatch = jest.fn()
      const expected = addNewUserThunk()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.addNewUser()

      expect(mockDispatch).toBeCalledWith(expected)
    })
  })
})