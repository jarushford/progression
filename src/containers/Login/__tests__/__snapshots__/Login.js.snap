// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login Component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
        >
          <h1>
            Log In
          </h1>
          <input
            name="loginEmail"
            onChange={[Function]}
            placeholder="email"
            type="email"
            value=""
          />
          <input
            name="loginPassword"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
          <button>
            Login
          </button>
        </form>,
        <form
          onSubmit={[Function]}
        >
          <h1>
            Sign Up
          </h1>
          <input
            name="signUpName"
            onChange={[Function]}
            placeholder="name"
            value=""
          />
          <input
            name="signUpEmail"
            onChange={[Function]}
            placeholder="email"
            type="email"
            value=""
          />
          <input
            name="signUpPassword"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
          <button>
            Sign Up
          </button>
        </form>,
      ],
      "className": "login-section",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Log In
            </h1>,
            <input
              name="loginEmail"
              onChange={[Function]}
              placeholder="email"
              type="email"
              value=""
            />,
            <input
              name="loginPassword"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />,
            <button>
              Login
            </button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Log In",
            },
            "ref": null,
            "rendered": "Log In",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "loginEmail",
              "onChange": [Function],
              "placeholder": "email",
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "loginPassword",
              "onChange": [Function],
              "placeholder": "password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login",
            },
            "ref": null,
            "rendered": "Login",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Sign Up
            </h1>,
            <input
              name="signUpName"
              onChange={[Function]}
              placeholder="name"
              value=""
            />,
            <input
              name="signUpEmail"
              onChange={[Function]}
              placeholder="email"
              type="email"
              value=""
            />,
            <input
              name="signUpPassword"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />,
            <button>
              Sign Up
            </button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign Up",
            },
            "ref": null,
            "rendered": "Sign Up",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "signUpName",
              "onChange": [Function],
              "placeholder": "name",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "signUpEmail",
              "onChange": [Function],
              "placeholder": "email",
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "signUpPassword",
              "onChange": [Function],
              "placeholder": "password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign Up",
            },
            "ref": null,
            "rendered": "Sign Up",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
          >
            <h1>
              Log In
            </h1>
            <input
              name="loginEmail"
              onChange={[Function]}
              placeholder="email"
              type="email"
              value=""
            />
            <input
              name="loginPassword"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />
            <button>
              Login
            </button>
          </form>,
          <form
            onSubmit={[Function]}
          >
            <h1>
              Sign Up
            </h1>
            <input
              name="signUpName"
              onChange={[Function]}
              placeholder="name"
              value=""
            />
            <input
              name="signUpEmail"
              onChange={[Function]}
              placeholder="email"
              type="email"
              value=""
            />
            <input
              name="signUpPassword"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />
            <button>
              Sign Up
            </button>
          </form>,
        ],
        "className": "login-section",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                Log In
              </h1>,
              <input
                name="loginEmail"
                onChange={[Function]}
                placeholder="email"
                type="email"
                value=""
              />,
              <input
                name="loginPassword"
                onChange={[Function]}
                placeholder="password"
                type="password"
                value=""
              />,
              <button>
                Login
              </button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Log In",
              },
              "ref": null,
              "rendered": "Log In",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "loginEmail",
                "onChange": [Function],
                "placeholder": "email",
                "type": "email",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "loginPassword",
                "onChange": [Function],
                "placeholder": "password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Login",
              },
              "ref": null,
              "rendered": "Login",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                Sign Up
              </h1>,
              <input
                name="signUpName"
                onChange={[Function]}
                placeholder="name"
                value=""
              />,
              <input
                name="signUpEmail"
                onChange={[Function]}
                placeholder="email"
                type="email"
                value=""
              />,
              <input
                name="signUpPassword"
                onChange={[Function]}
                placeholder="password"
                type="password"
                value=""
              />,
              <button>
                Sign Up
              </button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sign Up",
              },
              "ref": null,
              "rendered": "Sign Up",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "signUpName",
                "onChange": [Function],
                "placeholder": "name",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "signUpEmail",
                "onChange": [Function],
                "placeholder": "email",
                "type": "email",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "signUpPassword",
                "onChange": [Function],
                "placeholder": "password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sign Up",
              },
              "ref": null,
              "rendered": "Sign Up",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
